File: docs/aliyun/bailian-client.mjs
import OpenAI from "openai";

export const HOST =
  "https://ws-1pb723s50rmgu0wh.cn-beijing.maas.aliyuncs.com";
export const OPENAI_BASE_URL = `${HOST}/compatible-mode/v1`;
export const DASHSCOPE_BASE_URL = `${HOST}/api/v1`;

function key(apiKey) {
  const value = apiKey ?? process.env.DASHSCOPE_API_KEY;
  if (!value) throw new Error("Set DASHSCOPE_API_KEY or pass apiKey explicitly");
  return value;
}

export class BailianOpenAI {
  constructor({ apiKey, baseURL = OPENAI_BASE_URL } = {}) {
    this.client = new OpenAI({ apiKey: key(apiKey), baseURL });
  }

  chat({ model, messages, ...options }) {
    return this.client.chat.completions.create({ model, messages, ...options });
  }

  embedding({ model, input, ...options }) {
    return this.client.embeddings.create({ model, input, ...options });
  }
}

export class BailianDashScope {
  constructor({ apiKey, baseURL = DASHSCOPE_BASE_URL } = {}) {
    this.apiKey = key(apiKey);
    this.baseURL = baseURL.replace(/\/$/, "");
  }

  async request(path, body, { async = false } = {}) {
    const headers = {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };
    if (async) headers["X-DashScope-Async"] = "enable";
    const response = await fetch(`${this.baseURL}/${path.replace(/^\//, "")}`, {
      method: "POST",
      headers,
      body: JSON.stringify(body),
    });
    const payload = await response.json();
    if (!response.ok) {
      const error = new Error(payload.message || `HTTP ${response.status}`);
      error.payload = payload;
      throw error;
    }
    return payload;
  }

  text({ model, messages, parameters = {} }) {
    return this.request("services/aigc/text-generation/generation", {
      model,
      input: { messages },
      parameters: { result_format: "message", ...parameters },
    });
  }

  multimodal({ model, messages, parameters = {} }) {
    return this.request("services/aigc/multimodal-generation/generation", {
      model,
      input: { messages },
      parameters,
    });
  }

  embedding({ model, input, parameters = {} }) {
    return this.request("services/embeddings/text-embedding/text-embedding", {
      model,
      input: { texts: Array.isArray(input) ? input : [input] },
      parameters,
    });
  }

  image({ model = "wan2.7-image-pro", messages, parameters = {} }) {
    return this.request("services/aigc/multimodal-generation/generation", {
      model,
      input: { messages },
      parameters,
    });
  }

  video({ model, input, parameters = {} }) {
    return this.request(
      "services/aigc/video-generation/video-synthesis",
      { model, input, parameters },
      { async: true },
    );
  }
}

export async function discoverModels({ apiKey, pageSize = 200 } = {}) {
  const authorization = `Bearer ${key(apiKey)}`;
  const headers = { Authorization: authorization, "Content-Type": "application/json" };
  async function getAll(path, rowKey, extra, limit) {
    const rows = [];
    let page = 1;
    let last = {};
    while (true) {
      const params = new URLSearchParams({
        ...extra,
        page_no: String(page),
        page_size: String(limit),
      });
      const response = await fetch(`${DASHSCOPE_BASE_URL}/${path}?${params}`, { headers });
      const payload = await response.json();
      if (!response.ok) throw new Error(payload.message || `HTTP ${response.status}`);
      last = payload;
      const pageRows = payload.output?.[rowKey] ?? [];
      rows.push(...pageRows);
      const total = payload.output?.total;
      if (pageRows.length === 0 || (Number.isFinite(total) && rows.length >= total)) break;
      page += 1;
    }
    return {
      ...last,
      output: { ...(last.output ?? {}), [rowKey]: rows, total: last.output?.total ?? rows.length },
    };
  }

  const [catalog, permissions] = await Promise.all([
    getAll("models", "models", { supports: "inference" }, pageSize),
    getAll(
      "models/permissions",
      "permissions",
      { authorization_scope: "AUTHORIZED", action: "INFERENCE" },
      Math.min(pageSize, 200),
    ),
  ]);
  return { catalog, permissions };
}

File: docs/aliyun/bailian_client.py
"""Small, dependency-light clients for the Beijing Alibaba Cloud Model Studio workspace."""

from __future__ import annotations

import json
import os
from typing import Any
from urllib.parse import urlencode
from urllib.request import Request, urlopen


HOST = "https://ws-1pb723s50rmgu0wh.cn-beijing.maas.aliyuncs.com"
OPENAI_BASE_URL = f"{HOST}/compatible-mode/v1"
DASHSCOPE_BASE_URL = f"{HOST}/api/v1"


def _key(api_key: str | None) -> str:
    value = api_key or os.getenv("DASHSCOPE_API_KEY")
    if not value:
        raise RuntimeError("Set DASHSCOPE_API_KEY or pass api_key explicitly")
    return value


class BailianOpenAI:
    """OpenAI-compatible Chat and text-embedding client."""

    def __init__(self, api_key: str | None = None, base_url: str = OPENAI_BASE_URL):
        try:
            from openai import OpenAI
        except ImportError as exc:  # pragma: no cover - depends on user environment
            raise RuntimeError("Install the dependency with: pip install -U openai") from exc
        self.client = OpenAI(api_key=_key(api_key), base_url=base_url)

    def chat(self, *, model: str, messages: list[dict[str, Any]], **kwargs: Any) -> Any:
        return self.client.chat.completions.create(model=model, messages=messages, **kwargs)

    def embedding(self, *, model: str, input: str | list[str], **kwargs: Any) -> Any:
        return self.client.embeddings.create(model=model, input=input, **kwargs)


class BailianDashScope:
    """DashScope-native text and multimodal client."""

    def __init__(self, api_key: str | None = None, base_url: str = DASHSCOPE_BASE_URL):
        try:
            import dashscope
        except ImportError as exc:  # pragma: no cover - depends on user environment
            raise RuntimeError("Install the dependency with: pip install -U dashscope") from exc
        self.api_key = _key(api_key)
        dashscope.base_http_api_url = base_url.rstrip("/")
        self._dashscope = dashscope

    def text(self, *, model: str, messages: list[dict[str, Any]], **kwargs: Any) -> Any:
        return self._dashscope.Generation.call(
            api_key=self.api_key,
            model=model,
            messages=messages,
            result_format=kwargs.pop("result_format", "message"),
            **kwargs,
        )

    def multimodal(self, *, model: str, messages: list[dict[str, Any]], **kwargs: Any) -> Any:
        return self._dashscope.MultiModalConversation.call(
            api_key=self.api_key,
            model=model,
            messages=messages,
            **kwargs,
        )

    def embedding(self, *, model: str, input: str | list[str], **kwargs: Any) -> Any:
        return self._dashscope.TextEmbedding.call(
            api_key=self.api_key,
            model=model,
            input=input,
            **kwargs,
        )


def _get_json(path: str, *, api_key: str, query: dict[str, Any] | None = None) -> dict[str, Any]:
    url = f"{DASHSCOPE_BASE_URL.rstrip('/')}/{path.lstrip('/')}"
    if query:
        url = f"{url}?{urlencode(query, doseq=True)}"
    request = Request(
        url,
        headers={
            "Authorization": f"Bearer {_key(api_key)}",
            "Content-Type": "application/json",
        },
        method="GET",
    )
    with urlopen(request, timeout=30) as response:
        return json.load(response)


def _get_all(
    path: str,
    *,
    api_key: str,
    row_key: str,
    page_size: int,
    query: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Fetch every page while retaining the API response shape."""
    base_query = dict(query or {})
    page = 1
    combined: list[dict[str, Any]] = []
    last: dict[str, Any] = {}
    total: int | None = None
    while True:
        payload = _get_json(
            path,
            api_key=api_key,
            query={**base_query, "page_no": page, "page_size": page_size},
        )
        last = payload
        output = payload.get("output") or {}
        rows = output.get(row_key) or []
        combined.extend(rows)
        total = output.get("total", total)
        if not rows or (total is not None and len(combined) >= total):
            break
        page += 1
    output = dict(last.get("output") or {})
    output[row_key] = combined
    output["total"] = total if total is not None else len(combined)
    last["output"] = output
    return last


def discover_models(api_key: str | None = None, *, page_size: int = 200) -> dict[str, Any]:
    """Return catalog and current-workspace inference permissions without exposing the key."""
    key = _key(api_key)
    catalog = _get_all(
        "/models",
        api_key=key,
        row_key="models",
        page_size=page_size,
        query={"supports": "inference"},
    )
    permissions = _get_all(
        "/models/permissions",
        api_key=key,
        row_key="permissions",
        page_size=min(page_size, 200),
        query={
            "authorization_scope": "AUTHORIZED",
            "action": "INFERENCE",
        },
    )
    return {"catalog": catalog, "permissions": permissions}


def authorized_model_ids(data: dict[str, Any]) -> list[str]:
    """Extract models with inference=true from the permissions response."""
    rows = data.get("permissions", {}).get("output", {}).get("permissions", [])
    return [
        row["model"]
        for row in rows
        if row.get("model") and row.get("permissions", {}).get("inference") is True
    ]
